`construct-error.js` - child "error" fails because ["error" must be an object]

## Description

Introducing a query error on `MarkdownRemarkFrontmatter` seems to throw error with construct-error.js due to some error in validating an error type using the `joi` library. The error message thrown is:

`child "error" fails because ["error" must be an object]`

<details>
<summary>Original Build Error Log</summary>
This behavior will be removed in the next major release of gatsby-plugin-feed.
For more info, check out: https://gatsby.app/adding-rss-feed
success onPreBootstrap - 0.044 s
success source and transform nodes - 0.236 s
success building schema - 0.581 s
child "error" fails because ["error" must be an object]
Failed to validate error { ValidationError: child "error" fails because ["error" must be an object]
    at Object.exports.process (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/gatsby/node_modules/gatsby-cli/node_modules/@hapi/joi/lib/errors.js:203:19)
    at internals.Object._validateWithOptions (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/gatsby/node_modules/gatsby-cli/node_modules/@hapi/joi/lib/types/any/index.js:764:31)
    at module.exports.internals.Any.root.validate (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/gatsby/node_modules/gatsby-cli/node_modules/@hapi/joi/lib/index.js:145:23)
    at constructError (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/gatsby/node_modules/gatsby-cli/lib/structured-errors/construct-error.js:28:29)
    at Object.error (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/gatsby/node_modules/gatsby-cli/lib/reporter/index.js:100:29)
    at Object.panicOnBuild (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/gatsby/node_modules/gatsby-cli/lib/reporter/index.js:66:24)
    at Promise.catch.err (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/gatsby/dist/utils/api-runner-node.js:359:16)
    at tryCatcher (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/bluebird/js/release/promise.js:512:31)
    at Promise._settlePromise (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/bluebird/js/release/promise.js:569:18)
    at Promise._settlePromise0 (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/bluebird/js/release/promise.js:614:10)
    at Promise._settlePromises (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/bluebird/js/release/promise.js:690:18)
    at _drainQueueStep (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/bluebird/js/release/async.js:138:12)
    at _drainQueue (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/bluebird/js/release/async.js:131:9)
    at Async._drainQueues (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/bluebird/js/release/async.js:147:5)
    at Immediate.Async.drainQueues (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/bluebird/js/release/async.js:17:14)
  isJoi: true,
  name: 'ValidationError',
  details:
   [ { message: '"error" must be an object',
       path: [Array],
       type: 'object.base',
       context: [Object] } ],
  _object:
   { error: [ [Object] ],
     context: { sourceMessage: 'gatsby-node.js returned an error undefined' },
     text: 'gatsby-node.js returned an error undefined',
     level: 'ERROR',
     stack: [],
     docsUrl: 'https://gatsby.dev/issue-how-to' },
  annotate: [Function] }
⠋ createPages
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! unicorn-utterances-site@0.1.0 develop: `gatsby develop`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the unicorn-utterances-site@0.1.0 develop script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
</details>

Adding a console.error to the `contruct-error.js` file before doing the `joi` validation yeilds me a much better error message:

`GraphQLError: Cannot query field "attached" on type "MarkdownRemarkFrontmatter".`

<details>
<summary>The "Better" error</summary>
{ error:
   [ { GraphQLError: Cannot query field "attached" on type "MarkdownRemarkFrontmatter".
    at Object.Field (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/graphql/validation/rules/FieldsOnCorrectType.js:53:31)
    at Object.enter (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/graphql/language/visitor.js:324:29)
    at Object.enter (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/graphql/language/visitor.js:375:25)
    at visit (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/graphql/language/visitor.js:242:26)
    at validate (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/graphql/validation/validate.js:54:22)
    at graphqlImpl (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/graphql/graphql.js:99:50)
    at /Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/graphql/graphql.js:28:59
    at new Promise (<anonymous>)
    at graphql (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/graphql/graphql.js:26:10)
    at graphqlRunner (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/gatsby/dist/bootstrap/index.js:353:12)
    at Object.exports.createPages (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/gatsby-node.js:10:10)
    at runAPI (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/gatsby/dist/utils/api-runner-node.js:230:37)
    at resolve (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/gatsby/dist/utils/api-runner-node.js:352:15)
    at Promise._execute (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/bluebird/js/release/debuggability.js:313:9)
    at Promise._resolveFromExecutor (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/bluebird/js/release/promise.js:483:18)
    at new Promise (/Users/ccrutchley/git/Etc/UnicornUtter/blog-dev/node_modules/bluebird/js/release/promise.js:79:10)
       message: 'Cannot query field "attached" on type "MarkdownRemarkFrontmatter".',
       locations: [Array] } ],
  context: { sourceMessage: 'gatsby-node.js returned an error undefined' },
  text: 'gatsby-node.js returned an error undefined',
  level: 'ERROR',
  stack: [],
  docsUrl: 'https://gatsby.dev/issue-how-to' }
</details>

### Steps to reproduce

It's like 4AM, and I'm like falling asleep. Sincere apologies for not leaving a reproducible repo as a result :/ . If this is unable to be duplicated, whenever I wake up I can give it my best crack at one.

### Environment

Run `gatsby info --clipboard` in your project directory and paste the output here.
<details>
<summary>My system info</summary>
 ✘ 💅   ~/git/Etc/UnicornUtter/blog-dev   master ●✚ 
↳ npx gatsby info --clipboard

  System:
    OS: macOS High Sierra 10.13.6
    CPU: (4) x64 Intel(R) Core(TM) i5-7360U CPU @ 2.30GHz
    Shell: 5.3 - /bin/zsh
  Binaries:
    Node: 8.16.0 - ~/.nvm/versions/node/v8.16.0/bin/node
    Yarn: 1.17.0 - /usr/local/bin/yarn
    npm: 6.9.0 - ~/.nvm/versions/node/v8.16.0/bin/npm
  Languages:
    Python: 2.7.16 - /usr/local/bin/python
  Browsers:
    Chrome: 75.0.3770.100
    Safari: 12.1.1
  npmPackages:
    gatsby: ^2.11.3 => 2.11.3
    gatsby-image: ^2.2.3 => 2.2.3
    gatsby-plugin-feed: ^2.3.1 => 2.3.1
    gatsby-plugin-google-analytics: ^2.1.1 => 2.1.1
    gatsby-plugin-lunr: ^1.5.1 => 1.5.1
    gatsby-plugin-manifest: ^2.2.0 => 2.2.0
    gatsby-plugin-offline: ^2.2.0 => 2.2.0
    gatsby-plugin-prefetch-google-fonts: ^1.4.2 => 1.4.2
    gatsby-plugin-react-helmet: ^3.1.0 => 3.1.0
    gatsby-plugin-react-svg: ^2.1.1 => 2.1.1
    gatsby-plugin-sass: ^2.1.0 => 2.1.0
    gatsby-plugin-sharp: ^2.2.1 => 2.2.1
    gatsby-plugin-transition-link: ^1.12.4 => 1.12.4
    gatsby-remark-autolink-headers: ^2.1.0 => 2.1.0
    gatsby-remark-copy-linked-files: ^2.1.0 => 2.1.0
    gatsby-remark-images: ^3.1.2 => 3.1.2
    gatsby-remark-prismjs: ^3.3.0 => 3.3.0
    gatsby-remark-responsive-iframe: ^2.2.1 => 2.2.1
    gatsby-source-filesystem: ^2.1.1 => 2.1.1
    gatsby-transformer-json: ^2.2.0 => 2.2.0
    gatsby-transformer-remark: ^2.5.0 => 2.5.0
    gatsby-transformer-sharp: ^2.2.0 => 2.2.0

</details>

If you need another concrete example on recreating this error:

https://github.com/davidkartuzinski/openapprentice

I have simple instructions in the comments on top of the ./src/layout.js on two different components that, if restored, create this error.

I am developing on MacOS and `gatsby develop` produces a nice beautiful Gatsby site.  However, when pushed to netlify, with these two components I get the same error.    

PS.  This is new (within past 7 days or so)  I regularly spin up Gatsby sites (2-3 per week) and now (nearly) all my builds are failing - unless using just the Gatsby default starter.  

See below:

```
Failed to validate error { ValidationError: child "error" fails because ["error" must be an object]
11:03:58 PM:     at Object.exports.process (/opt/build/repo/node_modules/@hapi/joi/lib/errors.js:202:19)
11:03:58 PM:     at internals.Object._validateWithOptions (/opt/build/repo/node_modules/@hapi/joi/lib/types/any/index.js:764:31)
11:03:58 PM:     at module.exports.internals.Any.root.validate (/opt/build/repo/node_modules/@hapi/joi/lib/index.js:145:23)
11:03:58 PM:     at constructError (/opt/build/repo/node_modules/gatsby/node_modules/gatsby-cli/lib/structured-errors/construct-error.js:28:29)
11:03:58 PM:     at Object.error (/opt/build/repo/node_modules/gatsby/node_modules/gatsby-cli/lib/reporter/index.js:100:29)
11:03:58 PM:     at Object.panic (/opt/build/repo/node_modules/gatsby/node_modules/gatsby-cli/lib/reporter/index.js:58:24)
11:03:58 PM:     at reportFailure (/opt/build/repo/node_modules/gatsby/dist/commands/build.js:47:10)
11:03:58 PM:     at buildProductionBundle.catch.err (/opt/build/repo/node_modules/gatsby/dist/commands/build.js:102:5)
11:03:58 PM:   isJoi: true,
11:03:58 PM:   name: 'ValidationError',
11:03:58 PM:   details:
11:03:58 PM:    [ { message: '"error" must be an object',
11:03:58 PM:        path: [Array],
11:03:58 PM:        type: 'object.base',
11:03:58 PM:        context: [Object] } ],
11:03:58 PM:   _object:
11:03:58 PM:    { error:
11:03:58 PM:       [ './src/components/header.js\nModule not found: Error: Can\'t resolve \'./comingSoon\' in \'/opt/build/repo/src/components\'\nresolve \'./comingSoon\' in \'/opt/build/repo/src/components\'\n  using description file: /opt/build/repo/package.json (relative path: ./src/components)\n    Field \'browser\' doesn\'t contain a valid alias configuration\n    using description file: /opt/build/repo/package.json (relative path: ./src/components/comingSoon)\n      no extension\n        Field \'browser\' doesn\'t contain a valid alias configuration\n        /opt/build/repo/src/components/comingSoon doesn\'t exist\n      .mjs\n        Field \'browser\' doesn\'t contain a valid alias configuration\n        /opt/build/repo/src/components/comingSoon.mjs doesn\'t exist\n      .js\n        Field \'browser\' doesn\'t contain a valid alias configuration\n        /opt/build/repo/src/components/comingSoon.js doesn\'t exist\n      .jsx\n        Field \'browser\' doesn\'t contain a valid alias configuration\n        /opt/build/repo/src/components/comingSoon.jsx doesn\'t exist\n      .wasm\n        Field \'browser\' doesn\'t contain a valid alias configuration\n        /opt/build/repo/src/components/comingSoon.wasm doesn\'t exist\n      .json\n        Field \'browser\' doesn\'t contain a valid alias configuration\n        /opt/build/repo/src/components/comingSoon.json doesn\'t exist\n      as directory\n        /opt/build/repo/src/components/comingSoon doesn\'t exist\n[/opt/build/repo/src/components/comingSoon]\n[/opt/build/repo/src/components/comingSoon.mjs]\n[/opt/build/repo/src/components/comingSoon.js]\n[/opt/build/repo/src/components/comingSoon.jsx]\n[/opt/build/repo/src/components/comingSoon.wasm]\n[/opt/build/repo/src/components/comingSoon.json]\n @ ./src/components/header.js 3:0-38 6:65-75\n @ ./src/components/layout.js\n @ ./src/pages/404.js\n @ ./.cache/async-requires.js\n @ ./.cache/production-app.js',
11:03:58 PM:         './src/components/header.js\nModule not found: Error: Can\'t resolve \'./logo\' in \'/opt/build/repo/src/components\'\nresolve \'./logo\' in \'/opt/build/repo/src/components\'\n  using description file: /opt/build/repo/package.json (relative path: ./src/components)\n    Field \'browser\' doesn\'t contain a valid alias configuration\n    using description file: /opt/build/repo/package.json (relative path: ./src/components/logo)\n      no extension\n        Field \'browser\' doesn\'t contain a valid alias configuration\n        /opt/build/repo/src/components/logo doesn\'t exist\n      .mjs\n        Field \'browser\' doesn\'t contain a valid alias configuration\n        /opt/build/repo/src/components/logo.mjs doesn\'t exist\n      .js\n        Field \'browser\' doesn\'t contain a valid alias configuration\n        /opt/build/repo/src/components/logo.js doesn\'t exist\n      .jsx\n        Field \'browser\' doesn\'t contain a valid alias configuration\n        /opt/build/repo/src/components/logo.jsx doesn\'t exist\n      .wasm\n        Field \'browser\' doesn\'t contain a valid alias configuration\n        /opt/build/repo/src/components/logo.wasm doesn\'t exist\n      .json\n        Field \'browser\' doesn\'t contain a valid alias configuration\n        /opt/build/repo/src/components/logo.json doesn\'t exist\n      as directory\n        /opt/build/repo/src/components/logo doesn\'t exist\n[/opt/build/repo/src/components/logo]\n[/opt/build/repo/src/components/logo.mjs]\n[/opt/build/repo/src/components/logo.js]\n[/opt/build/repo/src/components/logo.jsx]\n[/opt/build/repo/src/components/logo.wasm]\n[/opt/build/repo/src/components/logo.json]\n @ ./src/components/header.js 2:0-26 6:104-108\n @ ./src/components/layout.js\n @ ./src/pages/404.js\n @ ./.cache/async-requires.js\n @ ./.cache/production-app.js' ],
11:03:58 PM:      context:
11:03:58 PM:       { sourceMessage: 'Generating JavaScript bundles failed undefined' },
11:03:58 PM:      text: 'Generating JavaScript bundles failed undefined',
11:03:58 PM:      level: 'ERROR',
11:03:58 PM:      stack: [],
11:03:58 PM:      docsUrl: 'https://gatsby.dev/issue-how-to' },
11:03:59 PM: failed during stage 'building site': Build script returned non-zero exit code: 1
11:03:58 PM:   annotate: [Function] }

```

--- a/packages/gatsby-cli/src/index.js
+++ b/packages/gatsby-cli/src/index.js
@@ -4,6 +4,7 @@
 // use require() with backtick strings so use the es6 syntax
 import "@babel/polyfill"
 const semver = require(`semver`)
+const util = require(`util`)
 
 const createCli = require(`./create-cli`)
 const report = require(`./reporter`)
@@ -24,10 +25,17 @@ if (!semver.satisfies(process.version, MIN_NODE_VERSION)) {
   )
 }
 
-process.on(`unhandledRejection`, error => {
+process.on(`unhandledRejection`, reason => {
   // This will exit the process in newer Node anyway so lets be consistent
   // across versions and crash
-  report.panic(`UNHANDLED REJECTION`, error)
+
+  // reason can be anything, it can be a message, an object, ANYTHING!
+  // we convert it to an error object so we don't crash on structured error validation
+  if (!(reason instanceof Error)) {
+    reason = new Error(util.format(reason))
+  }
+
+  report.panic(`UNHANDLED REJECTION`, reason)
 })
 
 process.on(`uncaughtException`, error => {

--- a/packages/gatsby-cli/src/structured-errors/error-map.js
+++ b/packages/gatsby-cli/src/structured-errors/error-map.js
@@ -76,6 +76,14 @@ const errorMap = {
     type: `CONFIG`,
     level: `ERROR`,
   },
+  "10126": {
+    text: context =>
+      `${context.pluginName} threw an error while running the ${
+        context.api
+      } lifecycle:\n\n${context.message}`,
+    type: `CONFIG`,
+    level: `ERROR`,
+  },
 }
 
 module.exports = { errorMap, defaultError: errorMap[``] }

--- a/packages/gatsby-cli/src/structured-errors/error-map.js
+++ b/packages/gatsby-cli/src/structured-errors/error-map.js
@@ -78,7 +78,7 @@ const errorMap = {
   },
   "11321": {
     text: context =>
-      `${context.pluginName} threw an error while running the ${
+      `"${context.pluginName}" threw an error while running the ${
         context.api
       } lifecycle:\n\n${context.message}`,
     type: `PLUGIN`,

--- a/packages/gatsby-cli/src/structured-errors/error-map.js
+++ b/packages/gatsby-cli/src/structured-errors/error-map.js
@@ -76,12 +76,12 @@ const errorMap = {
     type: `CONFIG`,
     level: `ERROR`,
   },
-  "10126": {
+  "11321": {
     text: context =>
       `${context.pluginName} threw an error while running the ${
         context.api
       } lifecycle:\n\n${context.message}`,
-    type: `CONFIG`,
+    type: `PLUGIN`,
     level: `ERROR`,
   },
 }

--- a/packages/gatsby-cli/src/structured-errors/error-schema.js
+++ b/packages/gatsby-cli/src/structured-errors/error-schema.js
@@ -10,7 +10,7 @@ const errorSchema = Joi.object().keys({
   text: Joi.string(),
   stack: Joi.array().items(Joi.object({}).unknown()),
   level: Joi.string().valid([`ERROR`, `WARNING`, `INFO`, `DEBUG`]),
-  type: Joi.string().valid([`GRAPHQL`, `CONFIG`]),
+  type: Joi.string().valid([`GRAPHQL`, `CONFIG`, `PLUGIN`]),
   filePath: Joi.string(),
   location: Joi.object({
     start: Position.required(),

--- a/packages/gatsby/src/utils/api-runner-node.js
+++ b/packages/gatsby/src/utils/api-runner-node.js
@@ -341,7 +341,7 @@ module.exports = async (api, args = {}, pluginSource) =>
         })
 
         reporter.panicOnBuild({
-          id: `10126`,
+          id: `11321`,
           context: {
             pluginName,
             api,

--- a/packages/gatsby/src/utils/api-runner-node.js
+++ b/packages/gatsby/src/utils/api-runner-node.js
@@ -333,7 +333,7 @@ module.exports = async (api, args = {}, pluginSource) =>
       let pluginName =
         plugin.name === `default-site-plugin`
           ? `gatsby-node.js`
-          : `Plugin ${plugin.name}`
+          : `${plugin.name}`
 
       return new Promise(resolve => {
         resolve(runAPI(plugin, api, { ...args, parentSpan: apiSpan }))
@@ -341,7 +341,24 @@ module.exports = async (api, args = {}, pluginSource) =>
         decorateEvent(`BUILD_PANIC`, {
           pluginName: `${plugin.name}@${plugin.version}`,
         })
-        reporter.panicOnBuild(`${pluginName} returned an error`, err)
+
+        let error = {}
+        if (err instanceof Error) {
+          error = {
+            error: err,
+          }
+        }
+
+        reporter.panicOnBuild({
+          ...error,
+          id: `10126`,
+          context: {
+            pluginName,
+            api,
+            message: err instanceof Error ? err.message : err,
+          },
+        })
+
         return null
       })
     }).then(results => {

--- a/packages/gatsby/src/utils/api-runner-node.js
+++ b/packages/gatsby/src/utils/api-runner-node.js
@@ -333,7 +333,7 @@ module.exports = async (api, args = {}, pluginSource) =>
       let pluginName =
         plugin.name === `default-site-plugin`
           ? `gatsby-node.js`
-          : `${plugin.name}`
+          : plugin.name
 
       return new Promise(resolve => {
         resolve(runAPI(plugin, api, { ...args, parentSpan: apiSpan }))

--- a/packages/gatsby/src/utils/api-runner-node.js
+++ b/packages/gatsby/src/utils/api-runner-node.js
@@ -331,9 +331,7 @@ module.exports = async (api, args = {}, pluginSource) =>
       }
 
       let pluginName =
-        plugin.name === `default-site-plugin`
-          ? `gatsby-node.js`
-          : plugin.name
+        plugin.name === `default-site-plugin` ? `gatsby-node.js` : plugin.name
 
       return new Promise(resolve => {
         resolve(runAPI(plugin, api, { ...args, parentSpan: apiSpan }))
@@ -342,21 +340,14 @@ module.exports = async (api, args = {}, pluginSource) =>
           pluginName: `${plugin.name}@${plugin.version}`,
         })
 
-        let error = {}
-        if (err instanceof Error) {
-          error = {
-            error: err,
-          }
-        }
-
         reporter.panicOnBuild({
-          ...error,
           id: `10126`,
           context: {
             pluginName,
             api,
             message: err instanceof Error ? err.message : err,
           },
+          error: err instanceof Error ? err : undefined,
         })
 
         return null

